metalearners/metalearner_grid_search_cv.py

```diff
@@ -0,0 +1,296 @@
+class _CVResult:
+    r"""Cross Validation Result."""
+
+    metalearner: MetaLearner
+    train_scores: dict
+    test_scores: dict
+    fit_time: float
+    score_time: float
+    cv_index: int
+
+
+def _fit_and_score(job: _FitAndScoreJob) -> _CVResult:
+    start_time = time.time()
+    job.metalearner.fit(job.X_train, job.y_train, job.w_train, **job.kwargs)
+    fit_time = time.time() - start_time
+
+    train_scores = job.metalearner.evaluate(
+        X=job.X_train,
+        y=job.y_train,
+        w=job.w_train,
+        is_oos=False,
+        scoring=job.scoring,
+    )
+    test_scores = job.metalearner.evaluate(
+        X=job.X_test,
+        y=job.y_test,
+        w=job.w_test,
+        is_oos=True,
+        oos_method=job.oos_method,
+        scoring=job.scoring,
+    )
+    score_time = time.time() - fit_time
+    return _CVResult(
+        metalearner=job.metalearner,
+        fit_time=fit_time,
+        score_time=score_time,
+        train_scores=train_scores,
+        test_scores=test_scores,
+        cv_index=job.cv_index,
+    )
+
+
+def _format_results(results: Sequence[_CVResult]):
+    rows = []
+    for result in results:
+        row: dict[str, str | int | float] = {}
+        row["metalearner"] = result.metalearner.__class__.__name__
+        nuisance_models = set(result.metalearner.nuisance_model_specifications().keys())
+        treatment_models = set(
+            result.metalearner.treatment_model_specifications().keys()
+        )
+        for model_kind in nuisance_models:
+            row[model_kind] = result.metalearner.nuisance_model_factory[
+                model_kind
+            ].__name__
+            for param, value in result.metalearner.nuisance_model_params[
+                model_kind
+            ].items():
+                row[f"{model_kind}_{param}"] = value
+        for model_kind in treatment_models:
+            row[model_kind] = result.metalearner.treatment_model_factory[
+                model_kind
+            ].__name__
+            for param, value in result.metalearner.treatment_model_params[
+                model_kind
+            ].items():
+                row[f"{model_kind}_{param}"] = value
+        row["cv_index"] = result.cv_index
+        row["fit_time"] = result.fit_time
+        row["score_time"] = result.score_time
+        for name, value in result.train_scores.items():
+            row[f"train_{name}"] = value
+        for name, value in result.test_scores.items():
+            row[f"test_{name}"] = value
+        rows.append(row)
+    df = pd.DataFrame(rows)
+    return df
+
+
+class MetaLearnerGridSearchCV:
+    """Exhaustive search over specified parameter values for a MetaLearner.
+
+    ``metalearner_params`` should contain the necessary params for the MetaLearner initialization
+    such as ``n_variants`` and ``is_classification``. It can also contain optional parameters
+    that all MetaLearners should be initialized with such as ``n_folds`` or ``feature_set``.
```

Comment:
```suggestion
    that all MetaLearners can be initialized with such as ``n_folds`` or ``feature_set``.
```
Are parameters truly **optional** of they **should** be provided?

Reply:
The mandatory ones for the `metalearner` need to be provided (`is_classification` and `n_variants`). The others are optional, what I meant with this `should` is not that the parameters have to be passed but if the metalearners **should** be initialized with some extra parameters, these can be passed through this dictionary.